### ISSUE TITLE
Control and Data Flow Analysis

### ISSUE THREAD

**Issue by justinethier** [justinethier/cyclone#205]
Both of these would enable a lot of optimizations, but CFA / DFA are complex to implement. I suppose the ideal would a 0-CFA that is linear in time instead of cubic (or worse!).

Some starting points are:

- http://matt.might.net/articles/implementation-of-kcfa-and-0cfa/
- http://cs.au.dk/~jmi/Midtgaard-CSur-final.pdf
- A Practical and Flexible Flow Analysis for Higher-Order Languages by Dybvig: https://www.cs.indiana.edu/~dyb/papers/toplas97.pdf

**Comment by tbmreza** [comment]
I'm looking to learn about CFA. Any writeups on how it's implemented at cyclone today?

**Comment by justinethier** [comment]
No, this is an area that could still be explored. Maybe the links above would be helpful, though.
